File: exercise/kotlin/day24/src/main/kotlin/domain/StockUnit.kt
package domain

import arrow.core.Either
import arrow.core.left
import arrow.core.right
import domain.core.Error

data class StockUnit(val value: Int) {

    companion object {
        fun from(stock: Int): Either<Error, StockUnit> =
            if (stock >= 0) StockUnit(stock).right()
            else Error("A stock unit cannot be negative").left()
    }

    fun isSupplied(): Boolean = value > 0
    fun decrease(): StockUnit = StockUnit(value - 1)
}

File: exercise/scala/day24/src/main/scala/domain/StockUnit.scala
package domain

import scala.util.{Either, Left, Right}
import domain.core.Error

case class StockUnit(value: Int) {
  def isSupplied: Boolean = value > 0
  def decrease: StockUnit = StockUnit(value - 1)
}

object StockUnit {
  def from(stock: Int): Either[Error, StockUnit] = {
    if (stock >= 0) Right(StockUnit(stock))
    else Left(Error("A stock unit cannot be negative"))
  }
}